Add Range Event Table to Bootstrap Schema

mrtracy opened the pull request:
This commit creates the 'rangelog.Event' table at bootstrap time. This table is
in the non-system reserved space; it is intended that the storage package will
write record to this table whenever a significant event occurs on a range (e.g.
a split or replica change).

This commit essentially just adds the table; it does not write any records to
the table, and the table's schema is incomplete. The number of testing changes
needed to support an additional table at startup were significant, so actual
usage of the table will be introduced in a subsequent commit.

Summary of necessary test changes:
- Many tests asserted that there would be a constant number of tables present on
  the server at boostrap table; in the case of some of the SQL logic tests, they
  assert very specifically on the expected set of initial tables.  All of these
  tests have been modified to account for the new tables, and where possible this
  has been done in a way that will be automatically accounted for if more tables
  are added in the future (using the new `GetBootstrapSchema()` function of the
  server package).
- Another category of tests asserted that there would be only a single _range_
  present on the server at startup; the addition of this new table causes the
  server to asychronously split the initial range as soon as the server starts up,
  which was causing these counts to be incorrect. The tests have been
  appropriately updated, in a way that will be automatically accounted for if more
  tables are added in the future.
- Another category of tests would create a TestServer, and then generate their
  own splits via the AdminSplit() function. However, the addition of the new
  table at startup means that a new TestServer will asynchronously attempt to
  split the initial range, and sometimes the calls to AdminSplit() would fail due
  to a transaction conflict, causing the tests to fail. To remedy this, these
  tests use the new function `TestServer.WaitForInitialSplits()` to wait for the
  initial splits to occur, eliminating the possibility of conflict.
- Finally, the asynchronous splitting of the initial range on TestServers was
  causing some tests to shut down incorrectly; specifically, there was an infinite
  retry loop that would cause the stopper to hang, eventually causing the test in
  question to time out. This was addressed by a combination of
  `WaitForInitialSplits()` the previous commit, which modified the server's
  DistSender to stop retrying if the server is stopping (specifically, once it is
  draining).

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3476)

<!-- Reviewable:end -->

Commit:
Add Range Event table to bootstrap schema

This commit creates the 'rangelog.Event' table at bootstrap time. This table is
in the non-system reserved space; it is intended that the storage package will
write record to this table whenever a significant event occurs on a range (e.g.
a split or replica change).

This commit essentially just adds the table; it does not write any records to
the table, and the table's schema is incomplete. The number of testing changes
needed to support an additional table at startup were significant, so actual
usage of the table will be introduced in a subsequent commit.

Summary of necessary test changes:

+ Many tests asserted that there would be a constant number of tables present on
the server at boostrap table; in the case of some of the SQL logic tests, they
assert very specifically on the expected set of initial tables.  All of these
tests have been modified to account for the new tables, and where possible this
has been done in a way that will be automatically accounted for if more tables
are added in the future (using the new `GetBootstrapSchema()` function of the
server package).

+ Another category of tests asserted that there would be only a single *range*
present on the server at startup; the addition of this new table causes the
server to asychronously split the initial range as soon as the server starts up,
which was causing these counts to be incorrect. The tests have been
appropriately updated, in a way that will be automatically accounted for if more
tables are added in the future.

+ Another category of tests would create a TestServer, and then generate their
own splits via the AdminSplit() function. However, the addition of the new
table at startup means that a new TestServer will asynchronously attempt to
split the initial range, and sometimes the calls to AdminSplit() would fail due
to a transaction conflict, causing the tests to fail. To remedy this, these
tests use the new function `TestServer.WaitForInitialSplits()` to wait for the
initial splits to occur, eliminating the possibility of conflict.

+ Finally, the asynchronous splitting of the initial range on TestServers was
causing some tests to shut down incorrectly; specifically, there was an infinite
retry loop that would cause the stopper to hang, eventually causing the test in
question to time out. This was addressed by a combination of
`WaitForInitialSplits()` the previous commit, which modified the server's
DistSender to stop retrying if the server is stopping (specifically, once it is
draining).

Commit:
DistSender stops retrying on server drain

This commit allows the DistSender to stop its retry loops in the case where the
local server's stopper is draining. This prevents certain infinite retry loops
which cause test cases to time out when a TestServer is stopped in the wrong
conditions.

To support this, this commit also creates the new Stopper.ShouldDrain() channel,
which allows goroutines to take action when a stopper enters draining mode.

Commit also adds LeakTest to the util/stop package.

bdarnell commented on kv/dist_sender_server_test.go:
```diff
@@ -83,9 +82,14 @@ func setupMultipleRanges(t *testing.T, splitAt ...string) (*server.TestServer, *
 	for _, key := range splitAt {
 		if err := db.AdminSplit(key); err != nil {
+			// Don't leak server goroutines.
+			defer s.Stop()
```
This is a weird place to defer a call to `s.Stop()` since it's far from the server creation. I'd call it without `defer` here.

bdarnell commented on server/intent_test.go:
```diff
@@ -104,9 +110,9 @@ func TestIntentResolution(t *testing.T) {
-			if err := s.db.AdminSplit(splitKey); err != nil {
-				t.Fatal(err)
-			}
+			util.SucceedsWithin(t, 500*time.Millisecond, func() error {
```
Why is this using `SucceedsWithin` instead of `WaitForInitialSplits`?

bdarnell commented on server/testserver.go:
```diff
@@ -196,6 +199,38 @@ func (ts *TestServer) StartWithStopper(stopper *stop.Stopper) error {
+// WaitForInitialSplits waits for the server to complete its expected initial
+// splits at startup. If the expected range count is not reached within the
+// supplied timeout, the current test will be aborted.
+func (ts *TestServer) WaitForInitialSplits(t util.Tester, timeout time.Duration) {
```
When would a caller pass in a different values for `timeout`? It seems like the timeout constant should be here rather than at each call site. 

Why isn't `WaitForInitialSplits` called from `StartTestServer`?

bdarnell commented on server/node.go:
```diff
@@ -92,6 +92,14 @@ func allocateStoreIDs(nodeID roachpb.NodeID, inc int64, db *client.DB) (roachpb.
 
+// GetBootstrapSchema returns the schema which will be used to bootstrap a new
+// server.
+func GetBootstrapSchema() sql.MetadataSchema {
```
This separation seems unnecessary; I'd rather just put all predefined tables in one place.

bdarnell commented on cli/cli_test.go:
```diff
@@ -210,60 +210,67 @@ func Example_ranges() {
 	c.Run("kv scan")
 	c.Run("quit")
 
-	// Output:
```
The comment `Output:` is special for example tests; it needs to be there to trigger the check that the output matches.

bdarnell commented:
LGTM

petermattis commented:
LGTM

---

Review status: 0 of 20 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kPT57G0ukFSQCC55C:-K5kPT57G0ukFSQCC55D:940241218) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L26)):*
Can you elaborate on what the plan is here? Having a table that contains `timestamp` and `rangeID` isn't terribly interesting. I was also imagining this would be a general purpose `eventlog` table for system events (e.g. cluster creation, node start, table creation, range splits / merges, etc). Given that ranges are described by a start and end key, how is that going to be mapped to an integer `rangeID`?

It's perfectly reasonable for this to be a work in progress, I'm just trying to understand the direction.

Edit: Ah, I see the commit message explains this a bit. Add a `TODO(mtracy)` here that the schema is incomplete and just for demonstration purposes at this point.

---

*[storage/log.go, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kQEZb8O9kzu6tLmc_:-K5kQEZb8O9kzu6tLmca:-223071475) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L27)):*
Any reason for this to not be in the `system` database (i.e. `system.rangelog` or `system.eventlog`)?

---

*[storage/log.go, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kPJNToQcish4u6nSc:-K5kPJNToQcish4u6nSd:-1502499022) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L29)):*
The column types do not line up which offends my eyes.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-:-K5kRQxr5ScDg80HRz17:676834233)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 20 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kRVDElqUjTKZTHGiW:-K5kRVDElqUjTKZTHGiX:3847840) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef292299c5f000f6cb52a14a3e8452c10379f7c9/cli/cli_test.go#L213)):*
This needs to be restored.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender_server_test.go, line 92 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcW-r1-92:-K5kecI9y_44uezrVE0T:-1336003056) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/kv/dist_sender_server_test.go#L92)):*
Why does this matter? The test is about to panic anyway, isn't it?

You changed `initReverseScanTestEnv` to take the server as an argument; can you make that change here, too? That would let you `defer s.Stop()` before calling this function.

---

*[server/intent_test.go, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kf9vOLf13ORT_QecQ:-K5kf9vOLf13ORT_QecR:2078231892) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L83)):*
Use (roachpb.Span).ContainsKey

---

*[server/server.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiWkiHY1zxGPrehNB:-K5kiWkiHY1zxGPrehNC:617819226) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/server.go#L130)):*
can you add a comment here?

---

*[server/server_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kfcR0ikzOQOczuE2Q:-K5kfcR0ikzOQOczuE2R:-1730431170) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/server_test.go#L155)):*
nit: make this a defer right after `s := ...`?

---

*[server/testserver.go, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kikUWDjbMiCs9PUJx:-K5kikUXDpgN3xyVlXhy:246221415) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L214)):*
This is fragile; what if the server is running in insecure mode?

---

*[sql/system_test.go, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kg3iuO-BXM3tmMGPQ:-K5kg3iuO-BXM3tmMGPR:1751712676) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/sql/system_test.go#L46)):*
add a period while you're here.

---

*[util/stop/stopper.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kgRPm42qP_qdV8k43:-K5kgRPm42qP_qdV8k44:-1223785292) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper.go#L66)):*
accidental comma

---

*[util/stop/stopper_test.go, line 303 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kgjtiVM8TaOZwS4Q0:-K5kgjtiVM8TaOZwS4Q1:-273553254) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L303)):*
Time to add leaktest to this package.

---

*[util/stop/stopper_test.go, line 320 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kh_nLpz7HE_O06c33:-K5kh_nLpz7HE_O06c34:-46675206) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L320)):*
can you make it clearer that this is the case because of the blocked RunWorker above?

in general, this test needs a few comments. it took me several minutes to figure it out.

---

*[util/stop/stopper_test.go, line 321 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiBiBxpNgMUmZ63L5:-K5kiBiCzYGVYDLIYyar:129503746) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L321)):*
I'm not sure you need the `time.After`s here; isn't a closed channel guaranteed to take precedence over `default`?

---

*[util/stop/stopper_test.go, line 322 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kgddqZU-yRH3VZCdQ:-K5kgddqZU-yRH3VZCdR:-1542324587) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L322)):*
remove this

---

*[util/stop/stopper_test.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiNBDD8FBno84TDxg:-K5kiNBDD8FBno84TDxh:-712120605) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L334)):*
I'd just block on `waiting` here and let the test timeout do its thing, but that's me.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on util/stop/stopper.go:
```diff
@@ -57,12 +57,13 @@ func (f CloserFn) Close() {
 	mu       sync.Mutex     // Protects the fields below
 	drain    *sync.Cond     // Conditional variable to wait for outstanding tasks
-	draining bool           // true when Stop() has been called
+	draining bool           // true when Stop() has been called,
```
?

tschottdorf commented on sql/split_test.go:
```diff
@@ -80,13 +80,10 @@ func TestSplitOnTableBoundaries(t *testing.T) {
-		t.Fatalf("expected no splits, but found %d ranges", num)
-	}
+	// The initial splits are happening on another thread, so it should succeed within a few
+	// milliseconds..
```
double dot

tschottdorf commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[server/intent_test.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5lhJVNphxK0zrIxrFk:-K5lhJVNphxK0zrIxrFl:-951810378) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L84)):*
also, `log.Infof` or remove this line?

---

*[util/stop/stopper_test.go, line 321 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiBiBxpNgMUmZ63L5:-K5lheiiBMJ1MQOkWh76:252425285) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L321)):*
I don't know that it would be guaranteed.

---

*[util/stop/stopper_test.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiNBDD8FBno84TDxg:-K5lhmizfgKLD4OIXfGI:-330895159) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L334)):*
same here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 7 of 21 files reviewed at latest revision, 22 unresolved discussions.

---

*[cli/cli_test.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kRVDElqUjTKZTHGiW:-K5lGMuVh9jcdAYxN5TL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef292299c5f000f6cb52a14a3e8452c10379f7c9/cli/cli_test.go#L213)):*
Done.

---

*[kv/dist_sender_server_test.go, line 92 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcW-r1-92:-K5lg5krdY_5EeFmmqqi:335376101) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/kv/dist_sender_server_test.go#L92)):*
Took tamir's suggestion, done.

---

*[server/intent_test.go, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kf9vOLf13ORT_QecQ:-K5ljvpj918ipM94nbyk:75720702) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L83)):*
I've added a TODO here, because roachpb.Span actually doesn't have a ContainsKey method; only RSpan has that method.  I do not know if that is deliberate or not (for some reason I feel that non-RKeys might be deliberately difficult to work with), either way this works for now and I don't want to do this as part of this PR.

---

*[server/intent_test.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5lhJVNphxK0zrIxrFk:-K5lkN0oyftgjTtObYxm:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L84)):*
Done.

---

*[server/intent_test.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcX-r1-113:-K5lH-MB1A9YdIEHaT72:-1235930526) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L113)):*
I missed one; before I created WaitForInitialSplits, I started using the pattern you see here in tests that needed to split.  That become unwieldy so I changed patterns, but apparently missed one here.

---

*[server/node.go, line 97 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcY-r1-97:-K5lHEX32jvVCB876fX4:-1309998530) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/node.go#L97)):*
Why should the sql package need to know about tables maintained and needed by other packages?

---

*[server/server.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiWkiHY1zxGPrehNB:-K5llXHCYk1vhyhXBXrj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/server.go#L130)):*
Done.

---

*[server/server_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kfcR0ikzOQOczuE2Q:-K5llf1B3Ib561VYd7sX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/server_test.go#L155)):*
Done.

---

*[server/testserver.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcb-r1-213:-K5llt5qtBtPhi967yrA:-1372424242) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L213)):*
I'm not sure why the timeout would be different in different places, it just seemed natural to expose the flexibility.  Maybe a test creates extra stuff during bootstrap, causing this to take longer.

I didn't put it in StartTestServer because I did not want to incur the slowdown cost for every test that runs a test server, but I suppose I could add it there if everyone is on board with that.

---

*[server/testserver.go, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kikUWDjbMiCs9PUJx:-K5llkT1TeIsfZWH0G2A:-1602574691) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L214)):*
Do any of our tests do this?  I don't know, I lifted this from another test that was getting range counts, assumed that all test servers were running secure by default.  Do you know how to make it less fragile?

---

*[sql/split_test.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPce-r1-84:-K5lmPrbp5jVy7hMkfem:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/sql/split_test.go#L84)):*
Done.

---

*[sql/system_test.go, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kg3iuO-BXM3tmMGPQ:-K5lmUEezDSOHuiUcogj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/sql/system_test.go#L46)):*
Done.

---

*[storage/log.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kPT57G0ukFSQCC55C:-K5mIC1pcxgmcsO8LlzU:-1352297949) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L26)):*
TODO added.

---

*[storage/log.go, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kQEZb8O9kzu6tLmc_:-K5lmouNwR3qjwSc7tqJ:-1778772529) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L27)):*
The data in the system database is gossiped and that seemed obviously inappropriate to me.  I assumed we would create multiple databases at startup; if everything is going into system, I imagine that would work, although I will have to refactor the MetadataSchema a bit to account for that.

---

*[storage/log.go, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kPJNToQcish4u6nSc:-K5ln9IVxV5l9Te7Ol2R:-1763357315) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L29)):*
Tab problem, fixed.

---

*[util/stop/stopper.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kgRPm42qP_qdV8k43:-K5loKb71e5XNM8jy6XS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper.go#L66)):*
Done.

---

*[util/stop/stopper.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPck-r1-66:-K5loKyagZZO2olrLmcS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper.go#L66)):*
Done.

---

*[util/stop/stopper_test.go, line 303 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kgjtiVM8TaOZwS4Q0:-K5lwiJpWUV7DmQ-iD_f:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L303)):*
Done.

---

*[util/stop/stopper_test.go, line 320 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kh_nLpz7HE_O06c33:-K5lscTSIrm4DzxnqoZ5:-512612970) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L320)):*
I have added a significant amount of annotation.

---

*[util/stop/stopper_test.go, line 321 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiBiBxpNgMUmZ63L5:-K5lsaBi4sbaTxZUU94d:-765502155) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L321)):*
I dunno, seems like it works.

---

*[util/stop/stopper_test.go, line 322 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kgddqZU-yRH3VZCdQ:-K5loWxOS-jUgzH_U-jA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L322)):*
Done.

---

*[util/stop/stopper_test.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiNBDD8FBno84TDxg:-K5lojz9V6ZHhqOqJbty:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L334)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 14 of 14 files at r2.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[server/intent_test.go, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kf9vOLf13ORT_QecQ:-K5mJTqklUcAXhcRnB75:-1352829979) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L83)):*
cc @tschottdorf

---

*[server/intent_test.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcX-r1-113:-K5mJZsweOMkVW1lBHd8:1322478076) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L113)):*
Seems you haven't fixed this one yet.

---

*[server/server.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiWkiHY1zxGPrehNB:-K5mJgLERCpFkUDhy8Aq:-920030677) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/server.go#L130)):*
Heh, can you expand the explanation in the comment? What is the cause of those infinite retries? The current comment doesn't tell me when this change would be safe to revert (or that it will never be).

---

*[server/testserver.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcb-r1-213:-K5mJx-1hX8aDe3nIBu5:-1214687885) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L213)):*
I'd be OK with that.

---

*[server/testserver.go, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kikUWDjbMiCs9PUJx:-K5mJzUKj24kSL6qE0IW:-430432275) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L214)):*
Yeah, use `ctx.RPCRequesScheme()`. There should be robust prior art somewhere.

---

*[server/testserver.go, line 237 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5mK58VfoAzrWNGcqV9:-K5mK58VfoAzrWNGcqVA:865272462) ([raw file](https://github.com/cockroachdb/cockroach/blob/da48295345c47b00e3d526b29c2921a595b8872e/server/testserver.go#L237)):*
s/has/had/, s/expect/expected/

nit: I'd also remove "server", since it's not helpful

---

*[util/stop/stopper_test.go, line 53 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5mKyMreOgCrz_8KHfs:-K5mKfG4OOBbNo7ljcJ_:1340958163) ([raw file](https://github.com/cockroachdb/cockroach/blob/da48295345c47b00e3d526b29c2921a595b8872e/util/stop/stopper_test.go#L53)):*
nit: kill this if you feel like it; this is the thing tobi and i commented on in your new test (couldn't see at tht time that it was mirrored here).

Come to think of it, I don't think you need both tests.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kQEZb8O9kzu6tLmc_:-K5mbx4EHlgH_aV51geO:394828606) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L27)):*
The system database is gossiped, but that isn't actually necessary. We need some of the existing tables in the system DB to be gossiped, but not all of them. I very much imagine we'll add more tables there that do not need to be gossiped.

The reason I want this rangelog/eventlog in the `system` DB is that it is a `system` table from a user's perspective. Whether the table is gossiped or not shouldn't affect which database it lives in.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[server/node.go, line 97 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcY-r1-97:-K5pgSe05DagTNq3uKnV:-950107605) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/node.go#L97)):*
I just feel like it would be simpler to have them all in one place; we may need to act on the predefined tables as a category in the future. In the end, the predefined tables belong to cockroach as a whole and not its subpackages. 

---

*[server/testserver.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcb-r1-213:-K5pi6aeQGihIvx3_2Cc:1280601459) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L213)):*
I think it's fine to put this in StartTestServer. Remember it's not sleeping for the given amount of time, that's just setting an upper bound, so it's fine if the time passed to `SucceedsWithin` is higher than it needs to be.

---

*[storage/log.go, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kQEZb8O9kzu6tLmc_:-K5pivSgk6LQyp5Ob3Eb:337129405) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L27)):*
I think it could be reasonable to have more than one "system" DB. Users don't care what's gossiped and what's not, but they may care about setting replication options, and I think we do have a couple of different classes of system data: the small metadata tables should be replicated widely, while log/timeseries data can be shunted off to spinning disks.  Using two "system" DBs can be a good way of classifying these tables for users so they don't have to think about them (or set zone configs for them) individually. (more than two system DBs starts to get complicated and we're probably better off sticking with tables in a single system db if we think there's that much variety). 

---

*[util/stop/stopper_test.go, line 321 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiBiBxpNgMUmZ63L5:-K5plX8X000cILUB0gb0:-1553969158) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L321)):*
Yes, this should be a `default`. The `default` case is guaranteed not to be chosen if any of the others can proceed without blocking

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

Commit:
Move WaitForInitialSplits() to testserver.Start()

WaitForInitialSplits() was previously being called at the start of several
specific tests. This commit moves it into TestServer.Start() so that its effects
are available for all tests.

The signature of WaitForInitialSplits has also been changed to no longer include
a timeout; instead, a pre-configured timeout of 1 second is always used.

Finally, WaitForInitialSplits now respects all server options correctly when
opening a client (security scheme, certs dir).

mrtracy commented:
Review status: 5 of 23 files reviewed at latest revision, 20 unresolved discussions.

---

*[server/intent_test.go, line 113 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcX-r1-113:-K7A9yb0d-ug6R74KD00:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/intent_test.go#L113)):*
Done.

---

*[server/node.go, line 97 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcY-r1-97:-K7ADMdqKsoP7kAbHtXl:-1027031556) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/node.go#L97)):*
We should discuss this with the group a bit, there seem to be differing opinions about where and how these tables should be defined.  For the purposes of this PR, i'm going to assume that it's fine as-is.

---

*[server/server.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiWkiHY1zxGPrehNB:-K7AA4eH02vqprf1Jn9j:-1490793840) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/server.go#L130)):*
Added a significant explanation of what causes the loop.

---

*[server/testserver.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcb-r1-213:-K7AIz02cS22Qj4K12si:1442167461) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L213)):*
I actually put it in `testserver.Start()`, which is more universal.

---

*[server/testserver.go, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kikUWDjbMiCs9PUJx:-K7A1eLgzrdBdVrv3Z9E:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L214)):*
Done.

---

*[server/testserver.go, line 237 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5mK58VfoAzrWNGcqV9:-K7A3O6tY5M7Y477-4WL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/da48295345c47b00e3d526b29c2921a595b8872e/server/testserver.go#L237)):*
Done.

---

*[storage/log.go, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kQEZb8O9kzu6tLmc_:-K7ABx45g5-Ibds45s-j:662203139) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/storage/log.go#L27)):*
There appear to be some differing opinions about this, we should likely discuss it in person a bit.  For the purposes of this PR, I'm going to assume it's fine as is (table is not in systemDB).

---

*[util/stop/stopper_test.go, line 321 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5kiBiBxpNgMUmZ63L5:-K7ADHs7RVhMElSHZQ6Z:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/util/stop/stopper_test.go#L321)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 5 of 23 files reviewed at latest revision, 20 unresolved discussions.

---

*[server/testserver.go, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476#-K5gnITn928s7B6AmPcb-r1-213:-K7CrEiQEsE_5eB6a8d-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e606de3eb96b687d35f6d4e5a020afae627ccf/server/testserver.go#L213)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3476)*
<!-- Sent from Reviewable.io -->
